services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java

```diff
@@ -1104,10 +1106,13 @@ public AccessTokenResponseBuilder generateRefreshToken() {
             boolean offlineTokenRequested = offlineAccessScope==null ? false : clientSessionCtx.getClientScopeIds().contains(offlineAccessScope.getId());
             generateRefreshToken(offlineTokenRequested);
             refreshToken.setScope(clientSessionCtx.getScopeString(true));
+            if (realm.isRevokeRefreshToken()) {
+                refreshToken.getOtherClaims().put(Constants.REUSE_ID, KeycloakModelUtils.generateId());
+            }
             return this;
         }
 
-        public AccessTokenResponseBuilder generateRefreshToken(String scope, AuthenticatedClientSessionModel clientSession) {
+        public AccessTokenResponseBuilder generateRefreshToken(String scope, AuthenticatedClientSessionModel clientSession, Object reuseId) {
```

Comment:
```suggestion
        public AccessTokenResponseBuilder generateRefreshToken(RefreshToken oldRefreshToken, AuthenticatedClientSessionModel clientSession) {
```

Maybe a nitpick, but is it possible to update signature of this method to something like the above? This can make it clearer that it is supposed to be used to generate refresh token from the old refresh token (both `scope` and `reuseId` could be retrieved from it).